### ISSUE TITLE
Paginated (>p1) hreflang link URLs are incorrect

### ISSUE THREAD

**Issue by chasegiunta** [nystudio107/craft-seomatic#342]
If visiting any paginated result page (/p2, /p3, etc) the hreflang meta links to other sites/languages are incorrect. Example:

```
<link href="https://www.applause.com/resources/all/p2" rel="canonical">
<link href="https://www.applause.com/" rel="home">
<link type="text/plain" href="https://www.applause.com/humans.txt" rel="author">
<link href="https://www.applause.com/es/resources/all/p2" rel="alternate" hreflang="es">
<link href="https://www.applause.com/fr/resources/all/p2" rel="alternate" hreflang="fr">
<link href="https://www.applause.com/de/resources/all/p2" rel="alternate" hreflang="de">
<link href="https://www.applause.com/resources/all/p2" rel="alternate" hreflang="en">
<link href="https://www.applause.com/resources/all/p2" rel="alternate" hreflang="x-default">
<link href="https://www.applause.com/resources/all" rel="prev">
<link href="https://www.applause.com/resources/all/p3" rel="next">
```

The hreflang links for es, fr, and de here should be `/es/recursos/all/p2`, `/fr/ressources/all/p2`, `/de/ressourcen/all/p2` respectively.

**Comment by khalwat** [comment]
It's asking Craft for the localized element URI for the given site tho?

https://github.com/nystudio107/craft-seomatic/blob/v3/src/helpers/DynamicMeta.php#L489

**Comment by chasegiunta** [comment]
@khalwat Yeah, dumping `$url` there actually returns the correct URL - it's must be getting changed later on?

```
            if ($matchedElement) {
                $url = $elements->getElementUriForSite($matchedElement->getId(), $site->id);
                var_dump('Site ID: '.$site->id);
                var_dump('URL: '.$url);
```
outputs:
```
string(10) "Site ID: 1"
string(18) "URL: resources/all"
string(10) "Site ID: 2"
string(19) "URL: ressourcen/all"
string(10) "Site ID: 3"
string(19) "URL: ressources/all"
string(10) "Site ID: 4"
string(17) "URL: recursos/all"
```
But the meta generated is still
```
<link href="https://applause.test/es/resources/all/p2" rel="alternate" hreflang="es">
<link href="https://applause.test/fr/resources/all/p2" rel="alternate" hreflang="fr">
<link href="https://applause.test/de/resources/all/p2" rel="alternate" hreflang="de">
<link href="https://applause.test/resources/all/p2" rel="alternate" hreflang="en">
<link href="https://applause.test/resources/all/p2" rel="alternate" hreflang="x-default">
```

**Comment by khalwat** [comment]
I'm not seeing anything in that code path that could change it; what if you `var_dump($localizedUrls)` just before the return?

```
        return $localizedUrls;
```

**Comment by chasegiunta** [comment]
It actually outputs the correct URLs there, too.
```     
  var_dump($localizedUrls);
  return $localizedUrls;
```
```
array(4) {
  [0]=>
  array(5) {
    ["id"]=>
    string(1) "1"
    ["language"]=>
    string(2) "en"
    ["ogLanguage"]=>
    string(2) "en"
    ["hreflangLanguage"]=>
    string(2) "en"
    ["url"]=>
    string(35) "https://applause.test/resources/all"
  }
  [1]=>
  array(5) {
    ["id"]=>
    string(1) "2"
    ["language"]=>
    string(2) "de"
    ["ogLanguage"]=>
    string(2) "de"
    ["hreflangLanguage"]=>
    string(2) "de"
    ["url"]=>
    string(39) "https://applause.test/de/ressourcen/all"
  }
  [2]=>
  array(5) {
    ["id"]=>
    string(1) "3"
    ["language"]=>
    string(2) "fr"
    ["ogLanguage"]=>
    string(2) "fr"
    ["hreflangLanguage"]=>
    string(2) "fr"
    ["url"]=>
    string(39) "https://applause.test/fr/ressources/all"
  }
  [3]=>
  array(5) {
    ["id"]=>
    string(1) "4"
    ["language"]=>
    string(2) "es"
    ["ogLanguage"]=>
    string(2) "es"
    ["hreflangLanguage"]=>
    string(2) "es"
    ["url"]=>
    string(37) "https://applause.test/es/recursos/all"
  }
}

```

Where is it appending the page trigger segments?

**Comment by khalwat** [comment]
@chasegiunta I think it may call that more than once per request, so delete your logs, make it happen, and then search the logs -- it's possible it's calling it more than once...

**Comment by chasegiunta** [comment]
What am I searching for in the logs? All of that above is using `var_dump()` so it's outputting those to the page on the request.

Have you tried to replicate this issue locally and haven't been able to do so?

**Comment by chasegiunta** [comment]
As far as I can tell, there's no seomatic logs generated in the `/storage/logs/` folder, even with dev mode enabled. I'm unable to narrow down the logs generated in web.log to anything useful for this case. 😕

**Comment by khalwat** [comment]
If you do `Craft::error(print_r($var, true), __METHOD__);` it'll add it to the `storage/logs/web.log`

**Comment by chasegiunta** [comment]
@khalwat Getting back on this. Yep! Looks like you're right. It's hitting that getLocalizedUrls() method twice on the request.

First time it's returning the non-paginated urls with the localized urls 
```
2019-03-11 12:25:00 [-][-][-][error][nystudio107\seomatic\helpers\DynamicMeta::getLocalizedUrls] Array
(
    [0] => Array
        (
            [id] => 1
            [language] => en
            [ogLanguage] => en
            [hreflangLanguage] => en
            [url] => https://applause.test/resources/all
        )

    [1] => Array
        (
            [id] => 2
            [language] => de
            [ogLanguage] => de
            [hreflangLanguage] => de
            [url] => https://applause.test/de/ressourcen/all
        )

    [2] => Array
        (
            [id] => 3
            [language] => fr
            [ogLanguage] => fr
            [hreflangLanguage] => fr
            [url] => https://applause.test/fr/ressources/all
        )

    [3] => Array
        (
            [id] => 4
            [language] => es
            [ogLanguage] => es
            [hreflangLanguage] => es
            [url] => https://applause.test/es/recursos/all
        )

)
```

Second time, which is what is output into the `<head>`

```
2019-03-11 12:25:02 [-][-][-][error][nystudio107\seomatic\helpers\DynamicMeta::getLocalizedUrls] Array
(
    [0] => Array
        (
            [id] => 1
            [language] => en
            [ogLanguage] => en
            [hreflangLanguage] => en
            [url] => https://applause.test/resources/all/p2
        )

    [1] => Array
        (
            [id] => 2
            [language] => de
            [ogLanguage] => de
            [hreflangLanguage] => de
            [url] => https://applause.test/de/resources/all/p2
        )

    [2] => Array
        (
            [id] => 3
            [language] => fr
            [ogLanguage] => fr
            [hreflangLanguage] => fr
            [url] => https://applause.test/fr/resources/all/p2
        )

    [3] => Array
        (
            [id] => 4
            [language] => es
            [ogLanguage] => es
            [hreflangLanguage] => es
            [url] => https://applause.test/es/resources/all/p2
        )

)
```

**Comment by chrisrowe** [comment]
@khalwat Were you able to reproduce? I'm seeing this too, looks like the method is being called twice. It's correct the first time but second time round it's using the default language slug across all locales.

**Comment by khalwat** [comment]
I'm going to attempt to duplicate it this week; I don't have a test environment that simulates this exactly (multi-site + paginated entries), so I'll have to create one.

If you're able, it'd be helpful to know where the second call is coming from?

**Comment by chrisrowe** [comment]
I didn't dig that far into it, @chasegiunta can you remember?

**Comment by chasegiunta** [comment]
I haven't dug into where that method is being called again - haven't been back on that specific project. Will do next week.

**Comment by khalwat** [comment]
So what exactly is `resources` here? It must be an index page of some kind?

What element is injected into this page?  e.g., where does the `ressourcen` come from, exactly?

**Comment by chrisrowe** [comment]
`/resources` is a listing template yeah, it's pulling from a channel. 
`resourcen` is the German translation 

<img width="756" alt="Screenshot 2019-03-27 at 16 32 41" src="https://user-images.githubusercontent.com/220456/55094269-fca8dc80-50ad-11e9-9559-4e5ef29dcea2.png">

**Comment by chrisrowe** [comment]
Just looked at this again, this is actually a single which loops over all entries. `ressourcen` is the German translation

![](http://up.chri.sr/Screen-Shot-2019-03-27-16-47-42.30-cGZbNQ.png)

**Comment by chasegiunta** [comment]
Echoing `{{ entry }}` on both `/all` & `/all/p2` template gives "Resources Listing" section.

**Comment by khalwat** [comment]
So ironically, this is caused by the change we made for your previously @chasegiunta where we changed it from `Craft::$app->getRequest()->getPathInfo()` to `Craft::$app->getRequest()->fullPath`

The former doesn't include any pagination in the URI, the latter does. The problem is that then the paginated URI doesn't match any elements, so it falls back on trying to do its best to create the URLs itself by using each site URL with the URI.

So what has to happen is we need to use the non-paginated `Craft::$app->getRequest()->getPathInfo()` and then manually re-paginate the URI. Which is a little tricky, because the page triggers can be done in a few different ways.
